### PR TITLE
Cell-centered Hypre and Petsc setup on GPU

### PR DESCRIPTION
**Pull request by WeiqunZhang**
## Summary

Do cell-centered hypre and Petsc solver setup on GPU.

## Checklist

The proposed changes:
- [ ] fix a bug or incorrect behavior in AMReX
- [x] add new capabilities to AMReX
- [ ] changes answers in the test suite to more than roundoff level
- [ ] are likely to significantly affect the results of downstream AMReX users
- [ ] are described in the proposed changes to the AMReX documentation, if appropriate

### CODE REVIEW

**Comment by sayerhs** [comment]
@WeiqunZhang I tested this code with AMR-Wind and the modifications to the cell-based solver seems to be working fine. There were two things I did when running this with AMR-Wind.

1. CMake – Add the necessary CUDA `cusparse` `curand` libraries to `amrex` so that the exe would link correctly. I see that you have already added that change to the GNUMakefile pathway. 

2. Add hypre GPU Init/Finalize calls so that it actually executes on GPUs. I show the timings below 

The code that needs to be added is shown below. I am not sure what is the right place to add this. Should this be the responsibility of application codes? Or can this be part of `amrex::Initialize() / amrex::Finalize()` calls. The latter would be transparent to applications that enable hypre and provide a consistent behavior. 

```C++
#if defined(AMREX_USE_CUDA) && defined(AMREX_USE_HYPRE)
  // hypre initialization when running on CUDA
  HYPRE_Init();
  hypre_HandleDefaultExecPolicy(hypre_handle()) = HYPRE_EXEC_DEVICE;
  hypre_HandleSpgemmUseCusparse(hypre_handle()) = 0;
#endif

#if defined(AMREX_USE_CUDA) && defined(AMREX_USE_HYPRE)
  // hypre finalize code
  HYPRE_Finalize();
#endif
```

**With hypre init/finalize**
```
WallClockTime: 1 Pre: 0.00035 Solve: 0.7311 Post: 0.00104 Total: 0.7324
WallClockTime: 2 Pre: 0.000355 Solve: 0.7263 Post: 0.00104 Total: 0.7277
WallClockTime: 3 Pre: 0.000337 Solve: 0.7493 Post: 0.00101 Total: 0.7507
WallClockTime: 4 Pre: 0.000342 Solve: 0.7551 Post: 0.00101 Total: 0.7565
WallClockTime: 5 Pre: 0.000339 Solve: 0.7553 Post: 0.231 Total: 0.9868
WallClockTime: 6 Pre: 0.000356 Solve: 0.7289 Post: 0.00102 Total: 0.7302
WallClockTime: 7 Pre: 0.000335 Solve: 0.7347 Post: 0.00103 Total: 0.7361
WallClockTime: 8 Pre: 0.00035 Solve: 0.7309 Post: 0.00101 Total: 0.7323
WallClockTime: 9 Pre: 0.000335 Solve: 0.7243 Post: 0.00104 Total: 0.7257
```

**without hypre init/finalize**
```
WallClockTime: 1 Pre: 0.00035 Solve: 1.747 Post: 0.00102 Total: 1.749
WallClockTime: 2 Pre: 0.000346 Solve: 1.747 Post: 0.00102 Total: 1.748
WallClockTime: 3 Pre: 0.000345 Solve: 1.712 Post: 0.00102 Total: 1.714
WallClockTime: 4 Pre: 0.000348 Solve: 1.752 Post: 0.00104 Total: 1.754
WallClockTime: 5 Pre: 0.000344 Solve: 1.754 Post: 0.244 Total: 1.998
WallClockTime: 6 Pre: 0.000352 Solve: 1.722 Post: 0.00101 Total: 1.724
WallClockTime: 7 Pre: 0.000343 Solve: 1.753 Post: 0.00105 Total: 1.755
WallClockTime: 8 Pre: 0.000347 Solve: 2.202 Post: 0.00102 Total: 2.203
WallClockTime: 9 Pre: 0.000352 Solve: 2.162 Post: 0.00102 Total: 2.164
```

**Timers with hypre init/finalize**
```
Name                                                              NCalls  Excl. Min  Excl. Avg  Excl. Max   Max %
-----------------------------------------------------------------------------------------------------------------
HypreABecLap3::solve()                                                13      5.374      5.374      5.374  47.33%
VisMF::Write(FabArray)                                                23      1.238      1.238      1.238  10.90%
HypreABecLap3::prepareSolver()                                        13     0.6852     0.6852     0.6852   6.04%
MLNodeLaplacian::Fsmooth()                                          2376     0.3914     0.3914     0.3914   3.45%
amr-wind::Temperature-Godunov::linsys_solve                           13     0.3746     0.3746     0.3746   3.30%
FillBoundary_nowait()                                              11630     0.3174     0.3174     0.3174   2.80%
MLTensorOp::apply()                                                  377     0.3154     0.3154     0.3154   2.78%
MLABecLaplacian::Fsmooth()                                          1872     0.2427     0.2427     0.2427   2.14%
MLCellLinOp::applyBC()                                              8812     0.2193     0.2193     0.2193   1.93%
```

**Timers without hypre init/finalize**
```
Name                                                              NCalls  Excl. Min  Excl. Avg  Excl. Max   Max %
-----------------------------------------------------------------------------------------------------------------
HypreABecLap3::prepareSolver()                                        13      14.39      14.39      14.39  55.23%
HypreABecLap3::solve()                                                13      6.265      6.265      6.265  24.04%
VisMF::Write(FabArray)                                                23      1.109      1.109      1.109   4.26%
MLTensorOp::apply()                                                  377     0.6177     0.6177     0.6177   2.37%
MLNodeLaplacian::Fsmooth()                                          2376      0.388      0.388      0.388   1.49%
amr-wind::Temperature-Godunov::linsys_solve                           13      0.372      0.372      0.372   1.43%
HypreABecLap3::loadVectors()                                          13     0.3363     0.3363     0.3363   1.29%
FillBoundary_nowait()                                              11630     0.3106     0.3106     0.3106   1.19%
MLABecLaplacian::Fsmooth()                                          1872     0.2426     0.2426     0.2426   0.93%
```

**Comment by sayerhs** [comment]
@WeiqunZhang I tried to use hypre for the Tensor solves and it complained that hypre solvers are only supported for `ncomp == 1`, is this expected or was it some issue with how I was setting up the run?

**Comment by WeiqunZhang** [comment]
I am not sure either whose responsibility it should be to call hypre gpu setup.

Currently, hypre is not set up for tensor solvers.

**Comment by WeiqunZhang** [comment]
Maybe we can add a ParmParse parameter for this.  amrex::Initialize will initialize hypre by default, but it can be turned off.

**Comment by WeiqunZhang** [comment]
For cmake, should we add anything for hypre to amrex?

**Comment by sayerhs** [comment]
@WeiqunZhang I like the idea of calling hypre init/finalize by default when CUDA is active and then allowing users to turn it off via input file option. 

For CMake, we will need to link cusparse/curand libraries when using a GPU-enabled hypre. Otherwise the executables will fail at link time. It seems best to directly add it to `amrex` target, otherwise the Plotfile executables will fail at link time.

**Comment by WeiqunZhang** [comment]
@zingale @maxpkatz `HYPRE_Init` and `HYPRE_Finalize` is now called by AMReX by default.  You might need to remove these calls from Castro.

File: Src/Base/AMReX.cpp [hunk 1]
```diff
@@ -538,16 +555,30 @@ amrex::Initialize (int& argc, char**& argv, bool build_parm_parse,
...
+#ifdef AMREX_USE_HYPRE
+    if (init_hypre) {
+        HYPRE_Init();
+#ifdef AMREX_USE_CUDA
+        hypre_HandleDefaultExecPolicy(hypre_handle()) = HYPRE_EXEC_DEVICE;
+        hypre_HandleSpgemmUseCusparse(hypre_handle()) = 0;
+#endif
+    }
+#endif
+
```
Comment (sayerhs) [inline Src/Base/AMReX.cpp]:
```suggestion
#if defined(AMREX_USE_HYPRE) && defined(HYPRE_USING_CUDA)
    if (init_hypre) {
        HYPRE_Init();
        hypre_HandleDefaultExecPolicy(hypre_handle()) = HYPRE_EXEC_DEVICE;
        hypre_HandleSpgemmUseCusparse(hypre_handle()) = 0;
    }
#endif
```

@WeiqunZhang We should probably account for the fact that the user might attempt to link to a hypre that is not compiled for CUDA. I had used something like the one above to handle situations where hypre and AMReX were built differently. Do you have any thoughts on that?

**Review by sayerhs — APPROVED** [review]

File: Src/Base/AMReX.cpp (hunk 1 above)
Comment (WeiqunZhang) [reply Src/Base/AMReX.cpp]: Thanks for the suggestion!  I have tested the current version using `HYPRE_USING_CUDA`.  It now works with hypre built with or without CUDA, when AMReX is built with CUDA.

**Pull request merged**